packages/api/amplify_api/lib/src/graphql/web_socket/services/web_socket_service.dart

```diff
@@ -0,0 +1,231 @@
+  /// Used to initialize a [WebSocketChannel] connection and return transformed
+  /// messages as [WebSocketEvent]
+  Stream<WebSocketEvent> init(
+    WebSocketState state,
+  );
+
+  /// Subscribe a [GraphQLRequest] inside the provided [WsSubscriptionBloc] to a open [WebSocketChannel]
+  Future<void> register(
+    ConnectedState state,
+    GraphQLRequest<Object?> request,
+  );
+
+  /// Unsubscribe a [GraphQLRequest] inside the provided [WsSubscriptionBloc]
+  Future<void> unsubscribe(
+    String subscriptionId,
+  );
+
+  /// Reconnect a [WebSocketChannel]
+  Future<void> reconnect();
+}
+
+/// {@template amplify_api.web_socket_service}
+/// Internal stateless [WebSocketService] implementation.
+/// {@endtemplate}
+class AmplifyWebSocketService
+    with AWSDebuggable, AmplifyLoggerMixin
+    implements WebSocketService, Closeable {
+  /// Service that contains the business logic for subscription management
+  AmplifyWebSocketService();
+
+  /// Sink used for sending messages to AppSync
+  @visibleForTesting
+  WebSocketSink? sink;
+
+  @override
+  Stream<WebSocketEvent> init(
+    WebSocketState state,
+  ) {
+    final sc = StreamCompleter<WebSocketEvent>();
+    _init(state, sc).catchError(sc.setError);
+
+    return sc.stream;
+  }
+
+  Future<void> _init(
+    WebSocketState state,
+    StreamCompleter<WebSocketEvent> sc,
+  ) async {
+    assert(
+      sink == null,
+      'Connection already exists, recreate service or clear sink',
+    );
+
+    try {
+      const webSocketProtocols = ['graphql-ws'];
+      final connectionUri = await generateConnectionUri(
+        state.config,
+        state.authProviderRepo,
+      );
+
+      final channel = WebSocketChannel.connect(
+        connectionUri,
+        protocols: webSocketProtocols,
+      );
+      sink = channel.sink;
+
+      final subscriptionStream = transformStream(channel.stream);
+
+      _send(WebSocketConnectionInitMessage());
+
+      return sc.setSourceStream(subscriptionStream);
+    } on Object catch (e, st) {
+      logger.error('Web socket error while initializing', e, st);
+      rethrow;
+    }
+  }
+
+  @override
+  Future<void> reconnect() {
+    // TODO(equartey): implement reconnect
+    throw UnimplementedError();
+  }
+
+  @override
+  Future<void> register(
+    ConnectedState state,
+    GraphQLRequest<Object?> request,
+  ) async {
+    // send registration msg
+    await _sendSubscriptionRegistrationMessage(
+      state,
+      request,
+    );
+  }
+
+  @override
+  Future<void> unsubscribe(
+    String subscriptionId,
+  ) async {
+    logger.debug('Attempting to cancel Operation $subscriptionId');
+    _send(
+      WebSocketStopMessage(id: subscriptionId),
+    );
+  }
+
+  void _send(WebSocketMessage message) {
+    try {
+      assert(sink != null, 'Sink must exist');
+      final msgJson = json.encode(message.toJson());
+      sink!.add(msgJson);
+    } on Exception catch (e) {
+      logger.warn(
```

Comment:
What is the goal of this? Catching exceptions here won't do anything since `sink.add` will only ever throw an Error

Reply:
Discussed offline, removed in favor of more allow for more visibility of errors.